compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/EmitCTypeConverter.cpp

```diff
@@ -65,6 +65,60 @@ EmitCTypeConverter::EmitCTypeConverter() {
   });
 }
 
+Type EmitCTypeConverter::convertTypeAsNonPointer(Type type) {
+  if (type.isa<IREE::VM::RefType>()) {
+    return emitc::OpaqueType::get(type.getContext(), "iree_vm_ref_t");
+  }
+
+  return convertType(type);
+}
+
+Type EmitCTypeConverter::convertTypeAsPointer(Type type) {
+  return emitc::PointerType::get(convertTypeAsNonPointer(type));
+}
+
+emitc::OpaqueType EmitCTypeConverter::convertTypeAsCType(Type type) {
+  Type convertedType = convertTypeAsNonPointer(type);
+  if (auto oType = convertedType.dyn_cast<emitc::OpaqueType>()) {
+    return oType;
+  }
+  if (auto iType = type.dyn_cast<IntegerType>()) {
+    std::string typeLiteral;
+    switch (iType.getWidth()) {
+      case 32: {
+        typeLiteral = "int32_t";
+        break;
+      }
+      case 64: {
+        typeLiteral = "int64_t";
+        break;
+      }
+      default:
+        assert(false && "unhandled type");
+    }
+    return emitc::OpaqueType::get(type.getContext(), typeLiteral);
+  }
+
+  if (auto fType = type.dyn_cast<FloatType>()) {
+    std::string typeLiteral;
+    switch (fType.getWidth()) {
+      case 32: {
+        typeLiteral = "float";
+        break;
+      }
+      case 64: {
+        typeLiteral = "double";
+        break;
+      }
+      default:
+        assert(false && "unhandled type");
+    }
+    return emitc::OpaqueType::get(type.getContext(), typeLiteral);
+  }
+
+  assert(false && "unhandled type");
```

Comment:
Asserts aren't flow control; in a release build these cases getting hit will result in obscure crashes somewhere up the call chain that tries to touch invalid memory as returned via a garbage register here. The asserts are fine if the expectation is that things are going to explode but in addition there should be some expected behavior defined: `return {};` or something. When it comes to a type converter returning empty is well-defined and will result in a proper error getting emitted to the user (likely a conversion failure and an unlegalized op using the unconvertable type) - way better than a stack trace that may not even have symbols :)

That is to say, design for predictable behavior for users with asserts disabled and then add asserts to make it easier for a developer to find the failure earlier when the expectation is that the developer is going to need to fix it.